Add support for generating ed25519 keys and certs

izolight opened the pull request:
It is now possible to generate ed25519 keys and certificates by specifying ed25519 as algo in the json.
This should fix #926

claucece commented:
Hi @lgarofalo ! I'll review this tomorrow ;)

claucece commented on helpers/helpers.go:
```diff
@@ -62,6 +63,8 @@ func KeyLength(key interface{}) int {
 	} else if rsaKey, ok := key.(*rsa.PublicKey); ok {
 		return rsaKey.N.BitLen()
+	} else if _, ok := key.(ed25519.PublicKey); ok {
+		return ed25519.PublicKeySize
```
Maybe here there is a way to use the size of the instantiation of the type rather than using the constant..

izolight replied:
ed25519 keys are always the same length, so the constant is ok

claucece commented on initca/initca.go:
```diff
@@ -197,6 +198,10 @@ func RenewFromSigner(ca *x509.Certificate, priv crypto.Signer) ([]byte, error) {
+	case ca.PublicKeyAlgorithm == x509.Ed25519:
+		if _, ok := priv.Public().(ed25519.PublicKey); !ok {
+			return nil, cferr.New(cferr.PrivateKeyError, cferr.KeyMismatch)
+		}
```
Why not compare here as well the `Public` with the `ca.Public`?

claucece commented on bundler/bundle.go:
```diff
@@ -119,6 +123,9 @@ func (b *Bundle) MarshalJSON() ([]byte, error) {
 		keyString = PemBlockToString(&pem.Block{Type: "EC PRIVATE KEY", Bytes: keyBytes})
+	case ed25519.PrivateKey:
+		keyBytes, _ = derhelpers.MarshalEd25519PrivateKey(key)
+		keyString = PemBlockToString(&pem.Block{Type: "PRIVATE KEY", Bytes: keyBytes})
```
Maybe something like Ed22519 PRIVATE KEY...

izolight replied:
i agree that it should be more meaningful, but i want to be compatible with the openssl output which generates it with just the PRIVATE KEY header
openssl genpkey -algorithm ed25519 -outform PEM -out test25519.pem

claucece replied:
I think openssl puts as a header 'PRIVATE KEY' to any algorithm. Just tried it with RSA.

claucece commented on initca/initca_test.go:
```diff
@@ -23,6 +23,7 @@ var validKeyParams = []csr.KeyRequest{
 	{A: "ecdsa", S: 256},
 	{A: "ecdsa", S: 384},
 	{A: "ecdsa", S: 521},
+	{A: "ed25519",},
```
What about putting here the size 256 as well?

claucece commented on initca/testdata/ca_csr_ed25519.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "key": {
+    "algo": "ed25519"
```
Should we add here the size as well? Maybe it is completely implied..

claucece commented on scan/crypto/tls/generate_cert.go:
```diff
@@ -51,6 +54,13 @@ func pemBlockForKey(priv interface{}) *pem.Block {
+			fmt.Fprintf(os.Stderr, "Unable to marshal ED25519 private key: %v", err)
+			os.Exit(2)
+		}
+		return &pem.Block{Type: "PRIVATE KEY", Bytes: b}
```
Probably change the type here as well ;)

claucece commented on transport/kp/key_provider.go:
```diff
@@ -210,6 +212,19 @@ func (sp *StandardProvider) Generate(algo string, size int) (err error) {
 		}
 		sp.internal.keyPEM = pem.EncodeToMemory(p)
 		sp.internal.priv = priv
+	case "ed25519":
```
Here the size might be passed as argument to the function as well.. so for consistency, what do you think of checking it?

claucece commented on transport/kp/key_provider.go:
```diff
@@ -210,6 +212,19 @@ func (sp *StandardProvider) Generate(algo string, size int) (err error) {
+			return err
+		}
+		p := &pem.Block{
+			Type:    "PRIVATE KEY",
```
Change the type as well..

claucece commented on transport/kp/key_provider.go:
```diff
@@ -319,6 +334,11 @@ func (sp *StandardProvider) Load() (err error) {
 			return
 		}
+	case *ed25519.PublicKey:
+		if p.Type != "ED25519 PRIVATE KEY" && p.Type != "PRIVATE KEY" {
```
Mmmm.. why are there two types here?

claucece commented:
Hi @lgarofalo !

I'll go ahead and make these little changes so it can be merged ;)

claucece commented:
Hi, @izolight ! I've created a new pull request with your changes, to fix some things. It is found here: https://github.com/cloudflare/cfssl/pull/1097. Let me know if it is ok ;)

izolight commented on bundler/bundle.go:
```diff
@@ -104,6 +106,8 @@ func (b *Bundle) MarshalJSON() ([]byte, error) {
 	case x509.ECDSA:
 		keyType = fmt.Sprintf("%d-bit ECDSA", keyLength)
+	case x509.Ed25519:
+		keyType = "ED25519"
```
no, was a typo :)

izolight commented on bundler/bundler.go:
```diff
@@ -581,6 +582,10 @@ func (b *Bundler) Bundle(certs []*x509.Certificate, key crypto.Signer, flavor Bu
+		case cert.PublicKeyAlgorithm == x509.Ed25519:
+			if _, ok := key.Public().(ed25519.PublicKey); !ok {
+				return nil, errors.New(errors.PrivateKeyError, errors.KeyMismatch)
+			}
```
i updated the pr to address this.

claucece replied:
Hey! The new PR is over here: https://github.com/cloudflare/cfssl/pull/1097, as it needed to incorporate the new commits for the master branch. We are also going to switch to use circl for ed25519, as if not we will have to remove go 1.12...

izolight commented on csr/csr.go:
```diff
@@ -72,6 +75,15 @@ func (kr *KeyRequest) Generate() (crypto.PrivateKey, error) {
 		}
 		return rsa.GenerateKey(rand.Reader, kr.Size())
+	case "ed25519":
+		if kr.Size() != 256 && kr.Size() != 0 {
```
If you omit the size parameter it will default to 0 on unmarshaling i think.
omitting it doesn't make sense for rsa or ecdsa, but ed25519 doesn't have parameters, so it looks cleaner to omit it in the json.

claucece replied:
`ed25519.SeedSize` return the size. Right now, seems to be checking for if the size is not 256 (which is correct) and not 0 (which is strange...)

claucece replied:
If it defaults to 0, then the condition should be an OR, I think..

izolight commented:
@claucece I've addressed some of your questions and updated the pr,
still a bit unsure about the type, maybe we should support PRIVATE KEY and Ed25519 PRIVATE KEY?

underrun commented:
any hope for this pr moving forward?

izolight commented:
merged changes from https://github.com/cloudflare/cfssl/pull/1097

lukevalenta commented on initca/initca_test.go:
```diff
@@ -46,19 +47,23 @@ var csrFiles = []string{
 	// Bad param
 	{A: "rsaCrypto", S: 2048},
 	{A: "ecdsa", S: 2000},
+	{A: "ed25519", S: 512},
```
The size parameter is ignored for ed25519, so this is not invalid. Let's remove this line.
```suggestion
```

Just to confirm, from your branch it looks like the test is indeed failing:
```
go test ./initca/...
[...]
--- FAIL: TestInvalidCryptoParams (0.00s)
    initca_test.go:240: InitCA with bad params should fail: <nil>
[...]
```

lukevalenta replied:
While we're at it, could you update the error message on lines 217 and 240? Right now they're checking that the error is `nil` and then printing it, but it would be more useful if they printed the invalid config instead.

Line 217:
```
if err == nil {
-    t.Fatal("InitCA with bad CAConfig should fail:", err)
+    t.Fatalf("InitCA with bad CAConfig should fail: %v", invalidCAConfig)
}
```

Line 240:
```
if err == nil {
-    t.Fatal("InitCA with bad params should fail:", err)
+    t.Fatalf("InitCA with bad params should fail: %v", invalidParam)
}
```

izolight replied:
updated it.

lukevalenta reviewed (changes requested):
Thanks for driving this PR forward! I left a few comments and requested changes, but overall this is looking good.

claucece reviewed (commented):
Looks good. I agree with @lukevalenta . 

I no longer work at Cloudflare, so I don't have time to check in depth PRs, but happy to see this work move forward.

izolight commented on initca/initca.go:
```diff
@@ -201,6 +202,10 @@ func RenewFromSigner(ca *x509.Certificate, priv crypto.Signer) ([]byte, error) {
 		if ca.PublicKey.(*rsa.PublicKey).N.Cmp(rsaPublicKey.N) != 0 {
 			return nil, cferr.New(cferr.PrivateKeyError, cferr.KeyMismatch)
 		}
+	case ca.PublicKeyAlgorithm == x509.Ed25519:
```
removed

izolight commented on initca/initca_test.go:
```diff
@@ -20,9 +20,10 @@ var validKeyParams = []csr.KeyRequest{
 	{A: "rsa", S: 3072},
 	{A: "rsa", S: 4096},
-	{A: "ecdsa", S: 256},
+	{A: "ecdsa"},
```
fixed

lukevalenta commented:
@nickysemenza this PR looks good to me but I lack permissions to approve. Would you be able to add this to your review queue?

adamacosta commented:
Is this going to make a release sometime soon? It was confusing as hell to spend the last two hours wondering why I couldn't generate a ed25519 key when your csr.go file pretty clearly shows a case match on it, to see the original request for support was from five years ago, then to finally look at the date on this merge and the date of the last release and realize you haven't released since April.

I'm perfectly willing to build from main and give it a whirl, but man, maybe merge the website updates from three and a half years ago: https://github.com/cloudflare/cfssl/issues/1095

Not a great experience to spend all this time digging through your source code because there's no documentation only to find out the source code doesn't actually match the latest release.

I know it's a free tool so honestly I apologize if this sounds harsh. Dealing with demanding jackasses on Github can be a pain, so honest to God, I've gotten a ton out of Cloudflare knowledge base material in general in the past and I really do appreciate it. This toolsuite is still a heck of a lot less inscrutable than openssl and at least the source code _is_ very readable.
